### PR TITLE
WOZ Metadata & crc32

### PR DESCRIPTION
**Pull request by hasseily**
Increased information available in pImageInfo:

- Added parsing and storing of WOZ metadata into the unordered map pImageInfo->umWOZMetadata.
- The full crc32 for any image is calculated and stored in pImageInfo->iCRC32.

### CODE REVIEW

**Comment by tomcw** [comment]
Hi Henri, thanks for the PR.

I assume this PR is spun out of PR #860 and feature #848.

Can you explain your use-case for this PR? (I can think of some, but would like to understand yours)

**Comment by hasseily** [comment]
Hello Tom,
you are correct I did some work earlier on this metadata and decided to expand it into a thorough PR.
My reasoning for this is as follows:

1. For the sake of completeness. The WOZ chunk parsing wasn't touching the metadata chunk. I added it. I'll have a follow-on PR to add volume name parsing from HDV and IMG.
2. AppleWin gains from knowing the name of the booted program. Displaying it in on the app window will be useful.
3. The largest repo of WOZ files is the "WOZ-a-day" by 4AM on archive.org. He has very valuable metadata in each WOZ, including `requires_machine`. For example, this field could have `2+|2e|2c|2gs` (pipe delimited). AppleWin could upon load of the image either notify the user that their chosen emulation model is not compatible with the image, or automatically switch model to run the image. The big plus here is for older programs that _don't_ run on the //e.
4. Of course for my remote control branch it's a necessity, telling a 3rd party app what is being run.
5. With the systematic crc calculation and storage in the image info, one could match it on certain signatures of known broken images, and notify the user upon load. I originally had AppleWin parse a blacklist file of signatures of broken (or not completeable) game images but dropped it for now.

**Comment by sicklittlemonkey** [comment]
For the sake of transparency and so we're all on the same page, @Michaelangel007 has strong opinions on IDing the running application by memory checks. Certainly this is more complicated and will no doubt require more manual curating of a DB of IDs, but I agree it's a more desirable option if feasible.

**Comment by tomcw** [comment]
I agree it's useful to read the WOZ meta-chunk data too, so I'm OK with taking this part of this PR.

But crc'ing the images files (disk or harddisk / plain, gzip or zip) doesn't scale across all images type, especially after images are written to.

**Comment by hasseily** [comment]
I also considered memory checks for IDing, but couldn't find truly unique memory areas consistent across programs. I did the code for it but upon testing I couldn't get unique consistent IDs.

I fully agree that the CRC32 as is has very limited value, but it will work as a first time check for downloaded images from asimov or archive.org. Just today for example someone was saying that the Under Fire DSK on Asimov wasn't working on AppleWin. I suggested a WOZ from archive.org, and that one worked fine, whoch leads me to believe that the problem isn't AppleWin but the DSK crack.

**Comment by tomcw** [comment]
This is a valid point. I have seen a few dodgy .DSK images myself (eg. Plato Spanish / #821).

So what you are suggesting is AppleWin maintaining a blacklist of image-CRCs so that the user can be warned when they try to boot them.

If that is the use-case, then I'm OK with including it in the PR. But I still think that more work needs to be done to compute the CRC for the unzipped (or un-gzip'ed) image, otherwise a CRC would be needed for zip & gzip images (and potentially zip could be done with different deflate algorithms or compression levels).

**Comment by hasseily** [comment]
Doesn't the PR already CRC the uncompressed image, whatever the compression method? I calculate the CRC on `pImageBuffer` itself. I must be mistaken in my understanding of the code then.

**Comment by tomcw** [comment]
You are right - you are CRC'ing the uncompressed image - so ignore my comment about that.

**Review by tomcw — CHANGES_REQUESTED** [review]
Can you change the code so that it's locally scoped within the META_CHUNK_ID case (see comment below attached to the code).

File: source/DiskImageHelper.cpp [hunk 1]
```diff
@@ -1483,6 +1485,16 @@ eDetectResult CWOZHelper::ProcessChunks(ImageInfo* pImageInfo, DWORD& dwOffset)
 	if (imageSizeRemaining < 0)
 		return eMismatch;
 
+	std::string szMetadata;
+	std::stringstream ssMetadata;
+	std::string szTmp;
+	const char cRowDelim = '\n';
+	std::string szColDelim = "\t";
+	size_t iTabPos;
+	std::unordered_map<std::string, std::string>::const_iterator pTitlePos;
```
[the remark this answers is not in the document]
Comment (hasseily) [reply source/DiskImageHelper.cpp]: Certainly, should have done that to start with. Done.

**Comment by hasseily** [comment]
The reason I hadn't done it was that no case statement was scoped and I thought it was one of the AppleWin standards. Fixed now.

**Comment by Michaelangel007** [comment]
@sicklittlemonkey It's not that I'm against a WOZ-based id, just that we _need_ a fallback plan for `.dsk`, `.do`, and `.po` where we don't have (any) meta-data. The only source of truth is the _actual_ program loaded into memory.

**Comment by hasseily** [comment]
You know better than me regarding RAM-based sigs. My first try was actually RAM-based but I couldn't find a reliable way to crc the correct memory area to make it unique enough but not so unique that it wouldn't trip on reboots.
One of the things I tried was to check RAM upon the first keystroke request. That's pretty nice but of course certain cracks make it impossible. For example there's a Ultima IV crack that expects a keystroke at the crack screen.

So I went back to do the woz metadata first, then I'll do the hdv/img volume name, and I'd be happy to help after that on a more generic solution.

**Comment by hasseily** [comment]
Let's close this as there's just no way to figure out what game is playing in a systematic manner. There are so many versions and revisions of everything, no chance. As qkumba said, it's akin to writing an antivirus signature checker.

**Comment by Michaelangel007** [comment]
[Almost] every game has a "main loop".  (Aside from Robotron which I won't get into.)

A game database could use this knowledge to build a "profile" of the main game loop, common routine address, "hot" instructions", multiple CRCs of RAM, to build an unique finger print.

As qkumba said, there are just WAY too many variations to do this based soley off of disk. There will always be false positives. Memory probing is  best chance you have of identifying which game is active.

**Pull request closed without merging**